### PR TITLE
Fixing issue with non-existant .Template when using tpl function

### LINKED ISSUE #10082: calling tpl: cannot retrieve Template.Basepath from values inside tpl function
Some commit in main broke `tpl` usage.

```
$ helm repo add grafana https://grafana.github.io/helm-charts
"grafana" has been added to your repositories

$ ~/helm-main/helm template grafana/grafana --debug
install.go:173: [debug] Original chart version: ""
install.go:190: [debug] CHART PATH: /home/foobar/.cache/helm/repository/grafana-6.16.2.tgz


Error: template: grafana/templates/deployment.yaml:36:28: executing "grafana/templates/deployment.yaml" at <include (print $.Template.BasePath "/configmap.yaml") .>: error calling include: template: grafana/templates/configmap.yaml:23:21: executing "grafana/templates/configmap.yaml" at <tpl $elemVal $>: error calling tpl: cannot retrieve Template.Basepath from values inside tpl function: console: "BasePath" is not a value
helm.go:88: [debug] template: grafana/templates/deployment.yaml:36:28: executing "grafana/templates/deployment.yaml" at <include (print $.Template.BasePath "/configmap.yaml") .>: error calling include: template: grafana/templates/configmap.yaml:23:21: executing "grafana/templates/configmap.yaml" at <tpl $elemVal $>: error calling tpl: cannot retrieve Template.Basepath from values inside tpl function: console: "BasePath" is not a value
```

The error comes from: https://github.com/grafana/helm-charts/blob/2d25705d20e439a51b5d7e3363bd79056036f3ea/charts/grafana/templates/configmap.yaml#L23

Works perfectly with 3.6.3.

Output of `helm version`:

- Working: `version.BuildInfo{Version:"v3.6.3", GitCommit:"d506314abfb5d21419df8c7e7e68012379db2354", GitTreeState:"clean", GoVersion:"go1.16.5"}`
- Broken: `version.BuildInfo{Version:"v3.6+unreleased", GitCommit:"dfec34e2dee69521d52d3ae753f8e6830d1cf477", GitTreeState:"clean", GoVersion:"go1.16.4"}`

Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.7", GitCommit:"132a687512d7fb058d0f5890f07d4121b3f0a2e2", GitTreeState:"clean", BuildDate:"2021-05-12T12:40:09Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.7", GitCommit:"132a687512d7fb058d0f5890f07d4121b3f0a2e2", GitTreeState:"clean", BuildDate:"2021-05-12T12:32:49Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
```

Some bisection on d506314abfb5d21419df8c7e7e68012379db2354...dfec34e2dee69521d52d3ae753f8e6830d1cf477 needed...

### PR DESCRIPTION
**Pull request by mattfarina**
This is a regression accidently introduced in #9957.

A delete call had been used on the Template key of vals. This caused
a condition where Template was not available when rendering via tpl.
The delete happened after ExecuteTemplate so the issue is surpsising.
It may possibly be a race condition. Existing tests did not catch it.
I tried to create a test that directly tested the issue and was
unable to replicate the error seen with real charts. This leads me
to believe it is a race condition in the underlying Go template
package.

The delete call was not there before #9957. It should be safe to
remove and keep that information.

Closes #10082

**What this PR does / why we need it**:

A regression was accidentally introduced. This fixes the regression.

**Special notes for your reviewer**:

The lack of tests is because this is difficult to test. Existing tests should have caught the issue and new tests failed to catch it. My current best guess is that this is due to a race condition somewhere in the Go template package.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Review by cpanato — APPROVED** [review]
/lgtm

but i defer to the maintainers

**Review by hickeyma — APPROVED** [review]
Thanks for this @mattfarina

**Comment by sathieu** [comment]
Thanks @mattfarina.

This race condition is scary.

**Comment by vflaux** [comment]
Hi,
I'm the one who introduced the regression (sorry).
I added the delete of the "Template" key because keeping it there exposes it in .Subcharts.xxx which makes no sense (but it's harmless).

What I didn't expect is for `renderWithReference` to be recursive. It it is called during the template execution by the "tpl" function.
When `renderWithReference` was called it deleted the "Template" key in the same map used by the caller template.
The error then occurs if the "Template" key was referenced in the following of the template.
`a: {{ $.Template.BasePath }}{{ tpl "" . }}` was ok
`a: {{ tpl "" . }}{{ $.Template.BasePath }}` was not

**Pull request merged**
